### PR TITLE
User Profile View

### LINKED ISSUE #321: Profile / follow
Basic profile page for sphere that allows you to follow sphere.

**Part of epic:** 
- https://github.com/subconsciousnetwork/subconscious/issues/330

**Follow ons:** 
- https://github.com/subconsciousnetwork/subconscious/issues/479
- https://github.com/subconsciousnetwork/subconscious/issues/478

### PR DESCRIPTION
**Pull request by bfollington**
Resolves #321 

Aims to introduce the visual components and get the navigation experience working. We'll wire this up to real data with backing services in #479.

# Demo
https://user-images.githubusercontent.com/5009316/229040041-a5568de6-3fff-4486-b33d-937f75ba0dd6.mov

# Features
- User profile view with scrollable columns for Recent Posts, Top Posts and Users Followed
- Integrated into the detail view, can transition between content and users freely
- Mapped out model and user actions
- DummyData utilities

# Tasks
- [x] Try metadata & noosphere block in meta sheet when omnibox is tapped
- [x] Design card components for notes & other users
    - [x] Backport to the address book as needed
- [x] Resolve best way to address a user's profile (including Omnibox rendering & link tapping)
- [x] Use correct types to build profile data (remove `Article`)
- [x] Follow / unfollow button on profile
- [x] Follow button is too big
- [x] Follow / unfollow user from meta sheet (make dynamic)
- [x] Metadata table on sheet (DID, CID, gateway URL)
- [x] Remove `AnyView` (revisit types of `TabbedColumnView`)
- [x] Move / clean up `Generatable` protocol & implementations
- [x] Tidy up props API interface, how does data flow in?
- [x] Stub out service layer to produce & request user profile data
- [x] Config flag + remove all hardcoded HACK and TODO

# Next Up
- https://github.com/subconsciousnetwork/subconscious/issues/479
- https://github.com/subconsciousnetwork/subconscious/issues/478

### CODE REVIEW

File: xcode/Subconscious/Shared/Components/Common/Byline/PetnameBylineView.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/Byline/PetnameBylineView.swift:47]: note to self: would be good to drop this if we can

File: xcode/Subconscious/Shared/Components/Common/ProfilePic.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/ProfilePic.swift:17]: opinion: Helps readability everywhere

File: xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/MemoEditorDetail.swift:56]: TODO: remove hard coded stub

File: xcode/Subconscious/Shared/Components/Detail/UserProfileDetailView.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Detail/UserProfileDetailView.swift:12]: Fix comment

File: xcode/Subconscious/Shared/Models/Slashlink.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Models/Slashlink.swift:67]: note to self: create Slug of profileSlug string and cache, avoids a `!` here

File: xcode/Subconscious/Shared/Components/AppView.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/AppView.swift:83]: Might hold this for an future PR

File: xcode/Subconscious/Shared/Components/Common/DidView.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/DidView.swift:31]: I realised this view has some problems with hit targets, even though it's not part of this feature.

File: xcode/Subconscious/Shared/Components/Common/Layout/TabbedThreeColumnView.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/Layout/TabbedThreeColumnView.swift:18]: This is the best way I found to get around `AnyView`

File: xcode/Subconscious/Shared/Components/Common/Layout/ThreeColumnView.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/Layout/ThreeColumnView.swift:38]: Seems repetitious but we may want to customise each of these in the future.

File: xcode/Subconscious/Shared/Components/Common/Profile/ProfilePic.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/Profile/ProfilePic.swift:17]: Subtle, but I think this helps readability a lot

File: xcode/Subconscious/Shared/Components/Common/Profile/UserProfileHeaderView.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/Profile/UserProfileHeaderView.swift:57]: Considering extracting to a function or sub-view because this is ugly. However it's likely placeholder anyway.

File: xcode/Subconscious/Shared/Components/Common/Tabs/TabButtonView.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/Tabs/TabButtonView.swift:25]: I could drop this, it's required for icon support in tabs but if we stick to plain text we don't need to worry about `@ViewBuilder`.

File: xcode/Subconscious/Shared/Components/FirstRun/FirstRunCreateSphereView.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/FirstRun/FirstRunCreateSphereView.swift:43]: Renamed, just noise

File: xcode/Subconscious/Shared/Config.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Config.swift:21]: This is the WIP table that shows IPFS hash, Gateway URL etc. on the profile metasheet

File: xcode/Subconscious/Shared/Library/DummyDataUtilities.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Library/DummyDataUtilities.swift:12]: I needed a way to populate all these views, I figure these could be useful to keep growing.

File: xcode/Subconscious/Shared/Models/Slashlink.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Models/Slashlink.swift:36]: This will be changed depending on the result of https://github.com/subconsciousnetwork/subconscious/issues/480

File: xcode/Subconscious/Shared/Models/Slashlink.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Models/Slashlink.swift:67]: This `!` cannot fail as long as this is static constant field with valid data.

File: xcode/Subconscious/Shared/Components/Common/Buttons/GhostPillButtonStyle.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/Buttons/GhostPillButtonStyle.swift:10]: Introduced to match the twitter follow vs following distinction:

image

File: xcode/Subconscious/Shared/Components/Common/Profile/UserProfileHeaderView.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/Profile/UserProfileHeaderView.swift:10]: This is the header:

image

File: xcode/Subconscious/Shared/Components/Common/Story/StoryEntryView.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/Story/StoryEntryView.swift:11]: Story entry:

image

File: xcode/Subconscious/Shared/Components/Common/Tabs/TabHeaderView.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Components/Common/Tabs/TabHeaderView.swift:15]: image

File: xcode/Subconscious/Shared/Config.swift
Comment (bfollington) [inline xcode/Subconscious/Shared/Config.swift:20]: Will set to false before merging, might be worth checking this brach out and test driving it with the flag on for review.

File: xcode/Subconscious/Shared/Components/Detail/UserProfileDetailMetaSheet.swift
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Detail/UserProfileDetailMetaSheet.swift:31]: Can you say more about the why of this?

File: xcode/Subconscious/Shared/Components/Detail/UserProfileDetailMetaSheet.swift
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Detail/UserProfileDetailMetaSheet.swift:180]: `&&`?

**Review by gordonbrander — CHANGES_REQUESTED** [review]

File: xcode/Subconscious/Shared/Components/OmniboxView.swift
Comment (bfollington) [reply xcode/Subconscious/Shared/Components/OmniboxView.swift:120]: When I added the user profile example we went over the limit for sibling views in SwiftUI and the compiler complains, we have multiple of this exact view in the list already so I dropped one copy.

File: xcode/Subconscious/Shared/Components/Detail/UserProfileDetailMetaSheet.swift
Comment (bfollington) [reply xcode/Subconscious/Shared/Components/Detail/UserProfileDetailMetaSheet.swift:180]: These can't actually be combined without losing functionality, there are three different cases here:

- Detail Section Hidden
- Detail Section Visible, Show Details Button Visible
- Detail Section Expanded, Hide Details Button Visible

File: xcode/Subconscious/Shared/Components/Detail/UserProfileDetailMetaSheet.swift
Comment (bfollington) [reply xcode/Subconscious/Shared/Components/Detail/UserProfileDetailMetaSheet.swift:31]: I want to make use of `ShareLink` as the zero-stress way to present a share sheet, but to style a `ShareLink` so that appears consistent with buttons in the `MetaTable` required jumping through some hoops. However, I came up with a better solution (I think) if you take another look.

File: xcode/Subconscious/Shared/Components/Common/Byline/PetnameBylineView.swift
Comment (bfollington) [reply xcode/Subconscious/Shared/Components/Common/Byline/PetnameBylineView.swift:32]: One was unused (removed), the other is a convenience to construct the view using a `Petname` without unpacking it every time.

File: xcode/Subconscious/Shared/Components/Common/Byline/SlashlinkBylineView.swift
Comment (bfollington) [reply xcode/Subconscious/Shared/Components/Common/Byline/SlashlinkBylineView.swift]: I went with `flatMap` (otherwise you get an option of an option and only unpack one layer). 

```swift
if let petname = petname.flatMap({ string in Petname(string) }) {
}
```


Sidenote: I've always found myself preferring `bind` as a name vs `flatMap` for Monads, except when it comes to sequences that is.

File: xcode/Subconscious/Shared/Components/Notebook/NotebookNavigationView.swift
Comment (bfollington) [reply xcode/Subconscious/Shared/Components/Notebook/NotebookNavigationView.swift:122]: I rolled this into the definition of `userProfile` in `Config`

File: xcode/Subconscious/Shared/Components/Common/Profile/UserProfileHeaderView.swift
Comment (bfollington) [reply xcode/Subconscious/Shared/Components/Common/Profile/UserProfileHeaderView.swift:10]: Updated to use the outline button in all cases 👍

File: xcode/Subconscious/Shared/Components/Detail/UserProfileDetailMetaSheet.swift
Comment (gordonbrander) [inline xcode/Subconscious/Shared/Components/Detail/UserProfileDetailMetaSheet.swift:19]: `.buttonStyle` works on `ShareLink`. So we can get rid of the wrapping button and use ShareLink directly.

```swift
ShareLink(item: item) {
    Label(
        label,
        systemImage: "square.and.arrow.up"
    )
}
.buttonStyle(RowButtonStyle())
```

File: xcode/Subconscious/Shared/Components/Detail/UserProfileDetailMetaSheet.swift
Comment (gordonbrander) [reply xcode/Subconscious/Shared/Components/Detail/UserProfileDetailMetaSheet.swift]: Comment on this below.

File: xcode/Subconscious/Shared/Components/Detail/UserProfileDetailMetaSheet.swift
Comment (bfollington) [reply xcode/Subconscious/Shared/Components/Detail/UserProfileDetailMetaSheet.swift:19]: I _swear_ I tried this before anything else and couldn't get it to behave it must've been a dream, it works fine now 🤦

**Pull request merged**
